### PR TITLE
Remove old Prices, updated for Gtk3

### PR DESCRIPTION
**Pull request by Bob-IT**
This pull request replaces number 143 as I still have not got to grips with rebasing and there was too much going on in the glade file changes. This encompasses all the changes that were requested in that pull request.

Bob

### CODE REVIEW

**Comment by gjanssens** [comment]
Oh, and don't worry about the conflicting files reported by github on the PR. I pulled the PR and it merged just fine. So as far as I'm concerned you can choose to either rebase your PR on current master (to pick up all the directory reshuffling I recently did) or continue to work on the PR as is. I can merge it either way.

**Comment by gjanssens** [comment]
One more detail, please check your PR for whitespace issues. There are a couple of lines adding trailing whitespace (or empty lines with whitespace).

File: src/gnome/dialog-price-edit-db.c [hunk 1]
```diff
@@ -216,53 +220,297 @@ gnc_prices_dialog_remove_clicked (GtkWidget *widget, gpointer data)
...
+        namespace_list = g_list_next (namespace_list);
+    }
+    g_list_free (commodity_list);
+    g_list_free (namespace_list);
+
+    return oldest_ts;
+}
+
+static GList *
+gnc_prices_dialog_get_commodity (GtkTreeView *view)
```
Comment (Bob-IT) [inline src/gnome/dialog-price-edit-db.c]: Should of spotted that one.

File: src/engine/gnc-pricedb.c [hunk 2]
```diff
@@ -1345,68 +1356,299 @@ pricedb_remove_foreach_pricelist (gpointer key,
...
+    Timespec time_b;
+    gnc_commodity *comma;
+    gnc_commodity *commb;
+    gnc_commodity *curra;
+    gnc_commodity *currb;
+    gint result;
+
+    if (!a && !b) return 0;
+    /* nothing is always less than something */
+    if (!a) return -1;
```
Comment (Bob-IT) [inline src/engine/gnc-pricedb.c]: Will add this so its complete.

File: src/engine/gnc-pricedb.c [hunk 3]
```diff
@@ -1345,68 +1356,299 @@ pricedb_remove_foreach_pricelist (gpointer key,
...
+        }
+        else
+            PINFO("## Remove price with date %s", date_buf);
+    }
+    else
+        PINFO("Keep price date is invalid");
+}
+
+static GNCPrice*
+save_cloned_price (GNCPrice *price, GNCPrice *clone_price)
```
Comment (Bob-IT) [inline src/engine/gnc-pricedb.c]: I considered some thing like this when first implementing but went the wrong way.

File: src/engine/gnc-pricedb.c [hunk 4]
```diff
@@ -1345,68 +1356,299 @@ pricedb_remove_foreach_pricelist (gpointer key,
...
+    GDateMonth month = g_date_get_month (date);
+
+    gint q = ((roundUp (22 - fiscal_start + month, 3)/3) % 4) + 1;
+
+    PINFO("Return fiscal quarter is %d", q);
+    return q;
+}
+
+static void
+gnc_pricedb_remove_old_prices_keep_last (GNCPriceDB *db, GDate *fiscal_end_date,
```
Comment (Bob-IT) [inline src/engine/gnc-pricedb.c]: will try to think of a different name.

**Comment by Bob-IT** [comment]
I am surprised about the white space, I thought I had a pre-commit hook that checked that, will have a look. One thing, I am using gnc_gdate_set_fiscal_year_end from app-utils/gnc-gdate-utils.c in the engine/gnc_pricedb.c, is it correct to add **#include "../app-utils/gnc-gdate-utils.h"** to gnc_pricedb.c or just copy the function in gnc_pricedb.c

Bob

**Comment by gjanssens** [comment]
No, you shouldn't make engine depend on app-utils. That would cause a circular dependency.

It looks like I moved gnc-gdate-utils too far. You can move it to engine.

**Comment by gjanssens** [comment]
And with that I mean you can move the full source and header of gnc-gdate-utils to engine, not just the function you need.

**Comment by jralls** [comment]
Or better still (in its own commit, of course) consolidate gnc-gdate-utils into gnc-date and adjust the #include directives accordingly.

**Comment by gjanssens** [comment]
I hesitated to propose that because gnc-gdate-utils is C api where I thought gnc-date has been converted to C++. Looking better at gnc-date it's still just C functions in a cpp file, so your suggestion is definitely better.

**Comment by Bob-IT** [comment]
I have made the changes required but not sure what happened with all the commits listed above but as you can see the new ones are at the bottom. If this can not easily be used I will scrap this PR and create a new one against the new directory structure.
Bob

**Comment by gjanssens** [comment]
Bob, I'm not sure what happened either. It looks like you did a backward rebase, that is you rebased master onto your branch and worked from there instead of rebasing your branch on master.
I think you can clean it up with an interactive rebase and dropping all my and @jralls' commits from your branch:

    git rebase -i 4f2d34f04328f676

(4f2d34f04328f676 is the base commit from where you branch starts). This will open an interactive text editor with a list of all commits on your branch. Delete all lines from there that are not your work, save and quit. Then git will recreate your branch based on the commits still in that list.

**Comment by jralls** [comment]
Bob, you need to rebase your prices-out2 on the current master, fixing conflicts  as you go, then force-push the result to your github repo. Once that's done the PR will be fixed, no need for a new one.

**Comment by Bob-IT** [comment]
OK after some poking around and using the command suggested along with some resets I think I have managed to get a workable set of changes.
Bob

**Comment by gjanssens** [comment]
Well done. I propose a few more interactive rebase excercises ;)

1. The last two commits should be merged into one (you can use "squash" or "fixup" in the interactive rebase list). This is so because the second to last commit won't even compile which we want to avoid as much as possible. Commits that don't build pollute bisect operations (an iterative process to find the first commit that shows a particular erroneous behaviour). This can't be avoided all the time, but in this case it's pretty straightforward. So please merge them into one commit.

2. Additionally, if you move the commits `Move gnc-gdate-utils.c/h from app-utils to engine`and `Add include for gnc-gdate-utils and remove copied function from earlier` higher up the list to right below `Temporary copy gnc_gdate_set_fiscal_year_end to gnc_pricdb.c` you can again squash these 3 commits together for a cleaner git history. Moving commits in the tree is a matter of cutting the line and pasting it somewhere else in the interactive rebase list.

**Comment by gjanssens** [comment]
Perfect. Thanks a lot!

**Pull request merged**